File: CIVITAS/src/civitas/MazoSorpresas.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package civitas;

import java.util.ArrayList;
import java.util.Collections;

/**
 *
 * @author Migue
 */
public class MazoSorpresas {
    private ArrayList<Sorpresa> sorpresas;
    private Boolean barajada;
    private int usadas;
    private Boolean debug;
    private  ArrayList<Sorpresa> cartasEspeciales;
    private Sorpresa ultimaSorpresa;
    
    private void init(){
        this.sorpresas = new ArrayList();
        this.cartasEspeciales = new ArrayList();
        this.barajada = false;
        this.usadas = 0;
        
    }
    
    MazoSorpresas (Boolean d){
        this.init();
        this.debug = d;
        if(d)
          Diario.getInstance().ocurreEvento("Se ha puesto el modo debug a "+d);
    }
    
    MazoSorpresas (){
        this.init();
        debug = false;
        
    }
    
    
    void alMazo(Sorpresa s){
        if(!barajada){
            this.sorpresas.add(s);
        }
    }
    
    Sorpresa siguiente(){
        if(!this.barajada || this.usadas == this.sorpresas.size()){
            if(!debug){
                this.usadas = 0;
                this.barajada = true; 
                Collections.shuffle(this.sorpresas); // Baraja el mazo
            }
           
        }
        
        this.usadas++;
        Sorpresa temp = this.sorpresas.get(0);
        this.sorpresas.remove(temp);
        this.sorpresas.add(temp);
        
        this.ultimaSorpresa = temp;
        
        return this.ultimaSorpresa;
        
    }
    
    void inhabilitarCartaEspecial (Sorpresa sorpresa){
        if(this.sorpresas.contains(sorpresa)){
            int index = this.sorpresas.indexOf(sorpresa);
            this.sorpresas.remove(index);
            this.cartasEspeciales.add(sorpresa);
            Diario.getInstance().ocurreEvento("Se ha cambiado una carta especial al mazo de especiales");
        }
    }
    
    void habilitarCataEspecial(Sorpresa sorpresa){
        if(this.cartasEspeciales.contains(sorpresa)){
            int index  = this.cartasEspeciales.indexOf(sorpresa);
            this.cartasEspeciales.remove(index);
            this.sorpresas.add(sorpresa);
             Diario.getInstance().ocurreEvento("Se ha cambiado una carta especial al mazo de sorpresas");
        }
        
    }
    
    
}

File: CIVITASRUBY/lib/mazo_sorpresas.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

require_relative "diario"
require_relative "sorpresa"

module Civitas
  class Mazo_sorpresas
    def initialize (d=false)
    
      init()
      @ultima_sorpresa = nil 
      @debug = d
      if(d)
        Diario.instance.ocurre_evento("se ha puesto el  modo debug del mazo a "+ d.to_s)
      end
      
    end
    
    def al_mazo(s)
      if (@barajada == false)
        @sorpresas.push(s)
      end
    end
    
    def siguiente
      if @barajada == false or @usadas == @sorpresas.length() 
        if @debug == false
          @usadas = 0
          @barajada = true

          @sorpresas = @sorpresas.shuffle
       
        end
      end
      
     @usadas += 1
     temp = @sorpresas[0]
     @sorpresas.delete(temp)
     @sorpresas.push(temp)
     
     @ultima_sorpresa = temp
     return @ultima_sorpresa
     
    end
    
    def inhabilitar_carta_especial (sorpresa)
     
      if @sorpresas.include? (sorpresa)
        @sorpresas.delete(sorpresa)
        @cartas_especiales.push(sorpresa)
            Diario.instance.ocurre_evento("Se ha cambiado una carta especial al mazo de especiales")
      end
    end
    
    def habilitar_carta_especial (sorpresa)
      if @cartas_especiales.include? (sorpresa)
        @cartas_especiales.delete(sorpresa)
        @sorpresas.push(sorpresa)
            Diario.instance.ocurre_evento("Se ha cambiado una carta especial al mazo de sorpresas")
      end
    end
    
    
    
    private 
    def init
 
      @sorpresas = Array.new
      @cartas_especiales = Array.new
      @barajada = false
      @usadas = 0
    end
  end
end